### PR TITLE
refactor(core): simplify array flatten logic

### PR DESCRIPTION
**Pull request by alan-agius4**
We can now use modern Javascript to get the same result.

### CODE REVIEW

File: packages/core/src/util/array_utils.ts [hunk 1]
```diff
@@ -48,23 +48,9 @@ export function arrayEquals<T>(a: T[], b: T[], identityAccessor?: (value: T) =>
...
-        dst = list.slice(0, i);
-      }
-      flatten(item, dst);
-    } else if (dst !== list) {
-      dst.push(item);
-    }
-  }
-  return dst;
+export function flatten(list: any[]): any[] {
+  // Cannot use `Infinity` as depth as `infinity` is not a number literal in TypeScript.
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply packages/core/src/util/array_utils.ts:52]: That still has the same problem and the return type is inferred incorrectly.

```ts
flatten(['a', ['b']]).map(c => c.replace('x', 'b'));
// Property 'replace' does not exist on type 'FlatArray<string | string[], 0 | 1 | -1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20>'.
```

**Review by dylhunn — APPROVED** [review]
reviewed-for: fw-core

**Comment by AndrewKushnir** [comment]
Just a quick comment (copy from a Slack message): it looks like this `flatten` function might be a part of the "hot path", i.e. invoked during change detection cycles, thus it's perf-sensitive. We'd need to double check that there are no perf implications (no execution time issues and ideally no extra arrays created).

**Comment by pkozlowski-opensource** [comment]
Also, there might be some logic specific to queries, I would like to make sure that we are not breaking things there

**Comment by alan-agius4** [comment]
From benchmarking as expected the native `flat` is faster with more than 50% more operations per second.





Test case name | Result
-- | --
ECMA .flat() | 779.8 Ops/sec 
Custom .flatten() | 332.2 Ops/sec


Screenshot 2022-12-06 at 09 08 00



https://www.measurethat.net/Benchmarks/ShowResult/362606

@pkozlowski-opensource, I am not following what you mean by specific logic for queries. 


Memory wise, it appears that it's almost the same.

**Comment by pkozlowski-opensource** [comment]
@alan-agius4 I was referring to [this usage of flatten in queries](https://github.com/angular/angular/blob/281c614113ec6106b73d3287c681e8e2911dd44c/packages/core/src/linker/query_list.ts#L160).

I was under the impression that we do depend on the result array identity change but the `arrayEquals` should take care of that, so we are probably good.

**Review by pkozlowski-opensource — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]
@alan-agius4 thanks for sharing perf results 👍

**Pull request closed without merging**